src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -755,6 +786,64 @@ ErrorIfMultiTaskRouterSelectQueryUnsupported(Query *query)
 
 
 /*
+ * TODO: update comments
+ */
+static void
+ErrorIfNotAllParticipatingTablesAreReferenceTables(Query *query)
+{
+	if (!AllParticipatingTablesAreReferenceTables(query))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("If data inserted into a reference table, "
+							   "all of the participating tables in the "
+							   "INSERT INTO ... SELECT query should be "
+							   "reference tables.")));
+	}
+}
+
+
+/*
+ * TODO: update comment
+ */
+bool
+AllParticipatingTablesAreReferenceTables(Query *query)
+{
```

Comment:
I'm not saying we *have* to do that, but we could very well integrate this into the pass already done via multi_relation_restriction_hook. RelationRestrictionContext already has hasDistributedRelation, we could easily (and cheaply!) add this knowledge, without additional passes over the parsetree.